pkg/gator/reader/read_constraints.go

```diff
@@ -207,3 +226,51 @@ func ReadK8sResources(r io.Reader) ([]*unstructured.Unstructured, error) {
 
 	return objs, nil
 }
+
+// ReadSyncRequirements parses the sync requirements from a
+// constraint template.
+func ReadSyncRequirements(t *templates.ConstraintTemplate) (*SyncRequirements, error) {
+	syncRequirements := &SyncRequirements{}
+	if t.ObjectMeta.Annotations != nil {
+		if annotation, exists := t.ObjectMeta.Annotations[SyncAnnotationName]; exists {
+			annotation = strings.Trim(annotation, "\n\"")
+			err := json.Unmarshal([]byte(annotation), syncRequirements)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return syncRequirements, nil
+}
+
+func (r *SyncRequirements) UnmarshalJSON(data []byte) error {
+	decoder := json.NewDecoder(bytes.NewReader(data))
+	decoder.DisallowUnknownFields()
+	var compactAnnotation [][]compactGVKEquivalentSet
+	err := decoder.Decode(&compactAnnotation)
+	if err != nil {
+		return err
+	}
+
+	*r = make([]map[schema.GroupVersionKind]struct{}, 0, len(compactAnnotation))
+	for _, compactRequirement := range compactAnnotation {
+		expandedRequirement := map[schema.GroupVersionKind]struct{}{}
+		for _, compactEquivalentSet := range compactRequirement {
+			expandCompactEquivalentSet(compactEquivalentSet, expandedRequirement)
+		}
+		*r = append(*r, expandedRequirement)
+	}
+	return nil
+}
+
+// Takes a compactGVKSet and expands and unions it with the set of
```

Comment:
nit: Having the usage of this function rely purely on pointer semantics makes it more confusing IMO.

To me it makes more sense to have a pure function which takes a compact representation and spits out a new data structure (the de-normalized representation).  This is easier to test and more resistant to coding error given that side-effects from previous calls to the function are not a factor in the output.

Reply:
Thanks! Changed